contracts/coconsensus/Coconsensus.sol

```diff
@@ -46,23 +47,230 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
-    mapping (bytes32 /* metachainId */ => ProtocoreI) protocores;
+    mapping (bytes32 /* metachainId */ => ProtocoreI) public protocores;
 
     /** Mapping of metachain id to the observers contract address. */
-    mapping (bytes32 /* metachainId */ => ObserverI) observers;
+    mapping (bytes32 /* metachainId */ => ObserverI) public observers;
 
     /** Mapping of metachain id to the domain separators. */
-    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) domainSeparators;
+    mapping (bytes32 /* metachainId */ => bytes32 /* domain separator */) public domainSeparators;
+
+
+    /* Special Functions */
+
+    /**
+     * @notice Setup function does the initialization of all the mosaic
+     *         contracts on the auxiliary chain.
+     *
+     * @dev This function can be called only once.
+     */
+    function setup() public {
+
+        require(
+            originMetachainId == bytes32(0),
+            "Coconsensus contract is already initialized."
+        );
+
+        originMetachainId = genesisOriginMetachainId;
+        auxiliaryMetachainId = genesisAuxiliaryMetachainId;
+
+        /*
+         * Setup self protocore contract first. All other protocores will
+         * require the dynasty from the self protocore to finalise the genesis
+         * checkpoint.
+         */
+        setupProtocores(auxiliaryMetachainId);
+
+        bytes32 currentMetachainId = genesisMetachainIds[SENTINEL_METACHAIN_ID];
+
+        // Loop through the genesis metachainId link list.
+        while (currentMetachainId != SENTINEL_METACHAIN_ID) {
+            /*
+             * The setup of self protocore is already done as a first step so
+             * skip the protocore setup if the current metachain id is equal to
+             * the auxiliary metachain id.
+             */
+            if (currentMetachainId != auxiliaryMetachainId) {
+                // Setup protocore contract for the given metachain id.
+                setupProtocores(currentMetachainId);
+            }
+
+            // Setup observer contract for the given metachain id.
+            setupObservers(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Anchor the state root in to the observer contracts.
+     *
+     * @param _metachainId Metachain id.
+     * @param _rlpBlockHeader RLP encoded block header.
+     *
+     * /pre `_metachainId` is not 0.
+     * /pre `_rlpBlockHeader` is not 0.
+     * /pre `blockHash` should exist in blockchains.
+     * /pre The dynasty of the reported block should be less than current dynasty.
+     * /pre The reported block should be finalized.
+     *
+     * /post Anchor the state root in the observer contract.
+     */
+    function observeBlock(
+        bytes32 _metachainId,
```

Comment:
Should we check that this is not self metachain id?

Reply:
done.